pgx-pg-sys/src/submodules/thread_check.rs

```diff
@@ -0,0 +1,82 @@
+//! Enforces thread-safety in `pgx`.
+//!
+//! It's UB to call into Postgres functions from multiple threads. We handle
+//! this by remembering the first thread to call into postgres functions, and
+//! panicking if we're called from a different thread.
+//!
+//! This is called from the current crate from inside the setjmp shim, as that
+//! code is *definitely* unsound to call in the presense of multiple threads.
+//!
+//! This is somewhat heavyhanded, and should be called from fewer places in the
+//! future...
+
+use std::num::NonZeroUsize;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+static ACTIVE_THREAD: AtomicUsize = AtomicUsize::new(0);
+#[track_caller]
+pub(crate) fn check_active_thread() {
+    let current_thread = nonzero_thread_id();
+    // Relaxed is sufficient as we're only interested in the effects on a single
+    // atomic variable, and don't need synchronization beyond that.
+    return match ACTIVE_THREAD.load(Ordering::Relaxed) {
+        0 => init_active_thread(current_thread),
+        thread_id => {
+            if current_thread.get() != thread_id {
+                thread_id_check_failed();
+            }
+        }
+    };
+}
+
+#[track_caller]
+fn init_active_thread(tid: NonZeroUsize) {
+    match ACTIVE_THREAD.compare_exchange(0, tid.get(), Ordering::Relaxed, Ordering::Relaxed) {
+        Ok(_) => unsafe {
+            // We won the race. Register an atfork handler to clear the atomic
+            // in any child processes we spawn.
+            extern "C" fn clear_in_child() {
+                ACTIVE_THREAD.store(0, Ordering::Relaxed);
+            }
+            libc::pthread_atfork(None, None, Some(clear_in_child));
+        },
+        Err(_) => {
+            thread_id_check_failed();
+        }
+    }
+}
+
+#[cold]
+#[inline(never)]
+#[track_caller]
+fn thread_id_check_failed() -> ! {
+    panic!("`pgx` may not not be used from multiple threads.");
+}
+
+fn nonzero_thread_id() -> NonZeroUsize {
+    // Returns the `addr()` of a thread local variable.
```

Comment:
wow.
cute.

Reply:
You don't think the `asm!` version was cuter? 🐱

Reply:
_wow_.